### ISSUE TITLE
explicit_auto_deref false positive

### ISSUE THREAD

**Issue by wdanilo** [rust-lang/rust-clippy#9763]
### Summary

explicit_auto_deref prints wrong warning

### Reproducer

I have a code:

```rust
let (shape, instance) = layer.instantiate(scene, &*self.data.borrow());
```

where `self.data` is of type `data: RefCell<S::ShapeData>`

where

```rust
pub fn instantiate<S>(
        &self,
        scene: &Scene,
        data: &S::ShapeData,
    ) -> (ShapeInstance<S>, LayerDynamicShapeInstance)
    where
        S: Shape,
    { ... }
```

and 

```rust
pub trait Shape: ... {
    type ShapeData: ...,
    ....
}
```

Clippy tells me:

```
warning: deref which would be done by auto-deref
   --> lib/rust/ensogl/core/src/gui/component.rs:155:58
    |
155 |         let (shape, instance) = layer.instantiate(scene, &*self.data.borrow());
    |                                                          ^^^^^^^^^^^^^^^^^^^^ help: try this: `&self.data.borrow()`
    |
    = note: `#[warn(clippy::explicit_auto_deref)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref

```

But it's incorrect. After changing the code according ot the suggestion, I've got:

```
error[E0271]: type mismatch resolving `<S as display::shape::primitive::system::Shape>::ShapeData == enso_prelude::Ref<'_, <S as display::shape::primitive::system::Shape>::ShapeData>`
   --> lib/rust/ensogl/core/src/gui/component.rs:155:39
    |
155 |         let (shape, instance) = layer.instantiate(scene, &self.data.borrow());
    |                                       ^^^^^^^^^^^ expected associated type, found struct `enso_prelude::Ref`
    |
    = note: expected associated type `<S as display::shape::primitive::system::Shape>::ShapeData`
                        found struct `enso_prelude::Ref<'_, <S as display::shape::primitive::system::Shape>::ShapeData>`
    = help: consider constraining the associated type `<S as display::shape::primitive::system::Shape>::ShapeData` to `enso_prelude::Ref<'_, <S as display::shape::primitive::system::Shape>::ShapeData>` or calling a method that returns `<S as display::shape::primitive::system::Shape>::ShapeData`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
```

### Version

```text
rustc 1.66.0-nightly (432abd86f 2022-09-20)
binary: rustc
commit-hash: 432abd86f231c908f6df3cdd779e83f35084be90
commit-date: 2022-09-20
host: aarch64-apple-darwin
release: 1.66.0-nightly
LLVM version: 15.0.0
```

**Comment by KenDJohnson** [comment]
Just ran into this as well. Seems to FP for a generic type with associated items. Simple example in [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=)

**Comment by lengyijun** [comment]
> Just ran into this as well. Seems to FP for a generic type with associated items. Simple example in [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=)

The link is broken

**Comment by Ralith** [comment]
I'm also getting false positives erroneously telling me to replace `*my_mutex.lock().unwrap()` with `my_mutex.lock().unwrap()`; unclear if it's the same exact bug.
